### PR TITLE
Docs/module descriptions

### PR DESCRIPTION
**Pull request by CalMacCQ**
Updated module descriptions to be more infromative adding links to the manual and notebook examples. Suggested by @rossduncan .

I've added an explanation of the offline circuit renderer (see https://github.com/CQCL/tket/issues/761 ) along with a rendered diagram using jupyter sphinx. I've added the `view_browser` function to the API docs as it was previously missing. I've also added some documentation notes for the offline circuit renderer.

Added module descriptions for `pytket.backends`, `pytket.placement`, `pytket.predictates` and `pytket.passes` with links to relevant examples and manual sections.

 Let me know if I should do the same for all of the remaining modules. There some that aren't well described like `pytket.tableau` and `pytket.partition`.

I've added a docs page for `pytket.wasm` as there wasn't one beforehand.

### CODE REVIEW

File: pytket/docs/predicates.rst [hunk 1]
```diff
@@ -1,5 +1,10 @@
 pytket.predicates
 ==================================
+
+In pytket predicates enforce properties of circuits. Each pytket :py:class:`Backend` has its own set of predicates which must be satisfied before a quantum circuit can be executed. There are predicates that enforce restrictions including gateset, number of qubits and classical control.
```
Comment (cqc-alec) [inline pytket/docs/predicates.rst:4]:
```suggestion
In pytket, predicates enforce properties of circuits. Each pytket :py:class:`Backend` has its own set of predicates which must be satisfied before a quantum circuit can be executed. There are predicates that enforce restrictions including gateset, number of qubits and classical control.
```

File: pytket/docs/requirements.txt [hunk 2]
```diff
@@ -1,4 +1,5 @@
 sphinx >= 4.5
 sphinx_autodoc_annotation >= 1.0
-sphinx_book_theme >= 0.3
+sphinx_book_theme == 0.3.3
```
Comment (cqc-alec) [inline pytket/docs/requirements.txt:3]:
```suggestion
sphinx_book_theme ~= 0.3.3
```

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: pytket/docs/backends.rst [hunk 3]
```diff
@@ -1,5 +1,21 @@
...
+2. `Comparing simulators <https://github.com/CQCL/pytket/blob/main/examples/comparing_simulators.ipynb>`_ - Discusses capabilities of simulator backends from ``pytket-qiskit``, ``pytket-pyquil``, ``pytket-qsharp``, ``pytket-projectq`` and ``pytket-qulacs``.
+3. `Qiskit integration <https://github.com/CQCL/pytket/blob/main/examples/qiskit_integration.ipynb>`_ - Covers compiling to IBM backends as well as using the `TketBackend <https://cqcl.github.io/pytket-qiskit/api/api.html#pytket.extensions.qiskit.tket_backend.TketBackend>`_ to run qiskit circuits.
+4. `Creating Backends <https://github.com/CQCL/pytket/blob/main/examples/creating_backends.ipynb>`_ - Tutorial showing how to define your own pytket :py:class:`Backend`. This could be used as a guide for developing new pytket extensions. 
+5. `Forest Portability example <https://github.com/CQCL/pytket/blob/main/examples/Forest_portability_example.ipynb>`_ - Tutorial on accessing Rigetti devices with the ``pytket-pyquil`` extension.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/backends.rst:14]: Thanks. Do you mean that I shouldn't add these notebook links from the API docs and just link to the examples README? 

Or should I just leave this as is and remove the pytket-pyquil links?

[the remark this answers is not in the document]
Comment (cqc-alec) [reply pytket/docs/backends.rst:14]: > Thanks. Do you mean that I shouldn't add these notebook links from the API docs and just link to the examples README?
> 
> Or should I just leave this as is and remove the pytket-pyquil links?

I think just link to the examples README, which is a better place for this info and will (hopefully!) get updated as the examples change.

File: pytket/docs/display.rst [hunk 4]
```diff
@@ -1,5 +1,18 @@
 pytket.circuit.display
 ==================================
 
+Conatins several functions for rendering interactive circuit diagrams.
+
+.. note:: Rendering circuits in jupyter notebooks with ``pytket.circuit.display`` requires an internet connection as the source code is in a `remote github repository <https://github.com/CQCL/pytket-circuit-renderer>`_ and is not installed with the pytket python package. Using the circuit renderer offline can be done by installing the `pytket-offline_display extension <https://github.com/CQCL/pytket-offline-renderer>`_.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/display.rst]: Done

File: pytket/docs/display.rst [hunk 5]
```diff
@@ -1,5 +1,18 @@
 pytket.circuit.display
 ==================================
 
+Conatins several functions for rendering interactive circuit diagrams.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/display.rst]: Done

File: pytket/docs/backends.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/backends.rst]: Okay, done

File: pytket/docs/backends.rst [hunk 6]
```diff
@@ -1,5 +1,21 @@
 pytket.backends
-==================================
+===============
+
+Contains :py:class:`Backend` abstract class and associated methods. In pytket a :py:class:`Backend` represents an interface between pytket and a quantum device or simulator. Different backends are defined in the various pytket extension modules and inherit from the core pytket :py:class:`Backend` class.
+
+Notebook tutorials on pytket backends
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/backends.rst]: I've removed the examples so haven't added a heading

File: pytket/docs/passes.rst [hunk 7]
```diff
@@ -1,6 +1,21 @@
 pytket.passes
 ==================================
 
+In pytket, compilation passes perform in-place transformations of circuits. From a user's point of view, passes are similar to `transforms <https://cqcl.github.io/tket/pytket/api/transform.html#>`_ expect passes allow for additonal predicate checking and compositionality.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/passes.rst]: Oops, done.

File: pytket/docs/passes.rst [hunk 8]
```diff
@@ -1,6 +1,21 @@
 pytket.passes
 ==================================
 
+In pytket, compilation passes perform in-place transformations of circuits. From a user's point of view, passes are similar to `transforms <https://cqcl.github.io/tket/pytket/api/transform.html#>`_ expect passes allow for additonal predicate checking and compositionality. 
+
+There are passes such as `FullPeepholeOptimise <https://cqcl.github.io/tket/pytket/api/passes.html#pytket.passes.FullPeepholeOptimise>`_ and  `KAKDecomposition <https://cqcl.github.io/tket/pytket/api/passes.html#pytket.passes.KAKDecomposition>`_ which are designed for general purpose circuit optimisation.
+
+Also there are special purpose passes such as `OptimisePhaseGadgets <https://cqcl.github.io/tket/pytket/api/passes.html#pytket.passes.OptimisePhaseGadgets>`_ and `PauliSimp <https://cqcl.github.io/tket/pytket/api/passes.html#pytket.passes.PauliSimp>`_ which perform optimisation by targetting phase gadget and Pauli gadget structures within circuits. For more on these optimisation techniques see the `corresponding publication <https://arxiv.org/abs/1906.01734>`_.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/passes.rst]: Done

File: pytket/docs/placement.rst [hunk 9]
```diff
@@ -1,5 +1,10 @@
 pytket.placement
 ==================================
+In order for the constraints of a :py:class:`Backend` to be solved we must first find assign device qubits to device independent (or program) qubits.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/placement.rst]: Done

File: pytket/docs/predicates.rst (hunk 1 above)
Comment (CalMacCQ) [reply pytket/docs/predicates.rst]: Done

File: pytket/docs/requirements.txt (hunk 2 above)
Comment (CalMacCQ) [reply pytket/docs/requirements.txt]: Done, will upgrade to the new release of the docs theme in a separate PR.

File: pytket/docs/display.rst [hunk 10]
```diff
@@ -1,5 +1,22 @@
 pytket.circuit.display
 ==================================
 
+Contains several functions for rendering interactive circuit diagrams.
+
+.. note:: Rendering circuits with ``pytket.circuit.display`` requires an internet connection. Using the pytket circuit renderer offline can be done by installing the `pytket-offline_display extension <https://github.com/CQCL/pytket-offline-renderer>`_.
```
Comment (cqc-alec) [inline pytket/docs/display.rst:6]:
```suggestion
.. note:: Rendering circuits with ``pytket.circuit.display`` requires an internet connection. Using the pytket circuit renderer offline can be done by installing the `pytket-offline-renderer extension <https://github.com/CQCL/pytket-offline-renderer>`_.
```

File: pytket/docs/display.rst [hunk 11]
```diff
@@ -1,5 +1,22 @@
...
+
+    ::
+
+        pip install pytket-offline_display
```
Comment (cqc-alec) [inline pytket/docs/display.rst:10]:
```suggestion
        pip install pytket-offline-renderer
```

**Review by cqc-alec — CHANGES_REQUESTED** [review]

**Comment by cqc-alec** [comment]
Note also the failure in the docs build.

**Comment by CalMacCQ** [comment]
Not sure why this run keeps failing

https://github.com/CQCL/tket/actions/runs/4377246155/jobs/7660740683

File: pytket/docs/display.rst [hunk 12]
```diff
@@ -1,5 +1,22 @@
...
+        pip install pytket-offline-renderer
+
 .. automodule:: pytket.circuit.display
-    :members: render_circuit_jupyter, render_circuit_as_html
+    :members: render_circuit_jupyter, render_circuit_as_html, view_browser
+
+.. jupyter-execute::
+
+    from pytket import Circuit
+    from pytket.circuit.display import render_circuit_jupyter
```
Comment (cqc-alec) [inline pytket/docs/display.rst:18]: Do you want to explain how to use the offline renderer too? (`from pytket.extensions.offline_display import render_circuit_jupyter`)

Comment (CalMacCQ) [reply pytket/docs/display.rst:18]: Ah yes I intended to do this but forgot to push the last commit I had locally. Will do that now.

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**